### PR TITLE
test(release): expand live workflow coverage and fix validation blockers

### PR DESCRIPTION
**Pull request by yingbull**
Adds seven UI-driven workflows for release 2026.08: episodes, diagnosis-triggered flowsheets, preventions, custom allergies, personal/professional contacts, consultation directory CRUD, and measurement history/plots. They exercise visible controls and verify persistence, reopening, cancellation, archive/deletion, opener refresh and owned-fixture cleanup. Shared audits detect startup errors, broken navigation, empty destinations, invalid PDFs and abandoned popups.

Based on and targeting `release/2026.08` at `fcb14db3c0bf568d10481dce144af526f80e4f4b`. No develop merge or Flyway/schema migration is included.

## Fixes required for reliable validation

- Repair SOAP startup injection, nullable messaging membership, vaccine catalogue fallback, measurement Plot rendering/encoding and the favicon route.
- Repair contact selection/serialization, new-row IDs, deletion, professional consent/status fields and duplicated control IDs.
- Validate contact ownership and planned reciprocal writes before mutations. Require related-patient write access only when creating a reverse row; preserve the original association and create a separate correctly typed reverse row without unrequested SDM/emergency flags.
- Preserve stored types for existing associations, including crafted or malformed submitted types. New associations still use their selected type. This prevents inconsistent relationship classification; it is not presented as an unauthorized related-patient-write finding.
- Keep fixture names within the database column limit. Complete verified database cleanup despite browser-teardown errors while still reporting those errors; retain the parent when child cleanup fails.
- Reject missing administration content panels and duplicate active prevention rows, which previously could pass their checks incorrectly. Preserve legitimate full-document navigation and validate its destination.

## Validation

- **Full Java suite: 12,139 tests, zero failures/errors, 51 skips.** All **30 contact cases** execute in normal discovery. The previously excluded legacy test filename was corrected and discovery verified in Surefire XML.
- **635 Node regressions** and **1,519 Python package tests** pass. Final documentation/manifest checks pass.
- The earlier corrected Ubuntu 26.04 DEB passes **seven workflows** and **14 login checks**, with credential/reset-field restoration verified. Later assertion changes also pass the live prevention workflow, **five administration destinations**, and **four real Chromium panel/navigation/error probes**.
- After the final type-preservation fix, the rebuilt DEB passes all six affected contact steps normally and with controlled request tampering. The old installed package fails the original-association type assertion after one tampered request; the final package preserves classification and reciprocity across two tampered saves. Unaffected workflows were not repeated on this last artifact.
- Negative Java cases fail before their fixes. Earlier browser/mutation probes also detect wrong reciprocal type/flags, missing persistence/opener refresh, lost startup errors, stale event listeners and silently ineffective cleanup.
- Installed contact class/JSP and measurement JSP hashes match the tested build. Final main DEB SHA-256: `537a1eb86d06495b57bf226ed36c0e0296743da25db49a5a4278a9a815828d27`. Reinstall applies **zero migrations**, preserving the release database's **23 successful migrations** and demo data; health checks pass.
- Final contact audit: **six owned patients, only 26 retained audit-log rows**. Earlier full-workflow audit: 12 owned patients, only 49 audit rows. Directory cleanup verifies zero leftovers. These are separate runs, not cumulative counts.
- VM: 6 GiB RAM, 2 GiB Java heap; sequential browsers. All host compilation/package generation occurs with the VM stopped. Final VM pass minimum available memory: **13.79 GiB host / 1.85 GiB guest**, zero swap/full memory PSI/guard interventions. Final full Java run peaks at 3.9 GiB under a 5 GiB cap. Temporary grants and original mounts are restored; **VM stopped normally**.

## Application findings retained for follow-up

The complete catalog and reproductions are in **#3682** and the [validation record](https://github.com/carlos-emr/carlos/blob/5542019d94185bf6a9a38a7b1ac797736f95d867/docs/ui-tests/release-2026.08-workflow-validation.md). Known application failures remain open:

- PDF labels/envelope, including Client Lab Label returning HTTP 200/PDF MIME with **zero bytes**.
- Three anonymous routes returning empty HTTP 200; this does not establish data disclosure.
- Missing calculator entry, Row Display CSRF input, Inbox HRM review-status partition mismatch, and provider-preference failures.
- Administration report errors and blank prevention lot-number search; Document Description Template's aborted fetch remains unclassified.
- **Internal patient contact Search displays an unavailable alert**, reproduced through the UI. The seeded existing-relationship fixture does not claim that search works.
- Source-only candidates: episode/scratchpad authorization, contact-ID namespace collisions, malformed contact numeric input before authorization, and changing an owned association's personal/professional category through the opposite submitted list. No live authorization bypass is claimed for these candidates.
- The existing DrugRef inactive-date issue remains open. Missing fixtures and external integrations are not counted as passing coverage.

The earlier 100-check sweep was 75 pass / 21 fail / 4 skip, with harness repairs during the sweep. Existing focused retests were 6 pass / 6 fail; administration opened 100 destinations with 20 repeated error observations across six destinations. The master-record audit retained four printing failures. This PR does not claim the whole application is defect-free; #3682 must remain open after merge.

All **20 commits have DCO sign-offs**. CodeRabbit reviewed final head `5542019d94` and found no new issues; Codex's earlier runtime review found no major issues. Unsupported review claims were answered with call-site or executable negative-test evidence. Sourcery declined the diff because it exceeds its 150,000-character limit. Final-head CI is green: **all reported checks complete, with zero failures or pending checks** (conditional jobs/reviewer skips and the neutral reviewer result remain explicitly reported). All review threads are resolved. Sonar’s quality gate passes; one pre-existing `getReverseRole` complexity warning remains (18 versus threshold 15). Comparing with the release base confirms identical helper logic after normalizing only the parameter rename; that existing maintainability debt is recorded in #3682. Required approving/code-owner review remains separate from automated checks; no merge was performed.

### CODE REVIEW

**Comment by yingbull** [comment]
@coderabbitai full review

File: scripts/lib/playwright-harness.js [hunk 1]
```diff
@@ -631,9 +633,7 @@ function wirePage(page, label, recorder, dialogHandler = null) {
   return wireStrictPage(page, label, recorder, {
     strictSignals: false,
     baseline: [],
-    dialogHandler: dialogHandler || (async (dialog) => {
-      await dialog.dismiss().catch(() => {});
-    }),
+    dialogHandler,
   });
```
[the remark this answers is not in the document]
Comment (yingbull) [reply scripts/lib/playwright-harness.js:637]: The fallback is still installed by wireStrictPage: its single dialog listener calls dialog.dismiss() when wiring.dialogHandler is null. Passing null from wirePage selects that fallback; it does not remove the listener. I strengthened the executable legacy regression to assert that dismiss is called. Custom legacy handlers retain ownership of recording, and strict handlers still retain every dialog. Focused and full Node tests pass.

File: src/main/webapp/WEB-INF/jsp/demographic/contactSearch.jsp [hunk 2]
```diff
@@ -178,7 +180,7 @@
             %>
             <tr bgcolor="<%=bgColor%>"
                 onMouseOver="this.style.cursor='hand';this.style.backgroundColor='pink';"
-                onMouseout="this.style.backgroundColor='<%=bgColor%>';" onClick="<carlos:encode value='<%= strOnClick %>' context="javaScriptAttribute"/>">
+                onMouseout="this.style.backgroundColor='<%=bgColor%>';" onClick="<carlos:encode value='<%= strOnClick %>' context="htmlAttribute"/>">
```
[the remark this answers is not in the document]
Comment (yingbull) [reply src/main/webapp/WEB-INF/jsp/demographic/contactSearch.jsp:192]: The dynamic contact name is already encoded for its JavaScript string literal with SafeEncode.forJavaScript before constructing strOnClick. The complete handler is then encoded for the outer HTML attribute. Encoding the complete executable handler as a JavaScript string escapes its syntax quotes and produced the live SyntaxError this fix addresses. The quoted-name live selection and executable serializer regression cover apostrophes, quotes and ampersands. Keeping the layered inner-JavaScript/outer-HTML encoding is intentional.

File: scripts/lib/playwright-ui.js [hunk 3]
```diff
@@ -121,7 +121,11 @@ async function clickOpensPopupOrNavigates(page, locator, options = {}) {
   // hide the outcome that actually occurred.
   const never = () => new Promise(() => {});
   const popupArrived = context.waitForEvent('page', { timeout })
-    .then((popup) => ({ page: popup, isPopup: true }), never);
+    .then((popup) => {
+      // The first document can execute before the opener's click resolves.
+      if (options.recorder) wireStrictPage(popup, label, options.recorder, options);
+      return { page: popup, isPopup: true };
+    }, never);
```
[the remark this answers is not in the document]
Comment (yingbull) [reply scripts/lib/playwright-ui.js]: Fixed in 7a112adfd0. The race now owns on/off event listeners and removes both immediately when an outcome wins, times out, or the click fails. Popup recording occurs synchronously in the event callback. Executable EventEmitter regressions verify listener removal, later-popup isolation, same-address/main-frame navigation, first-document errors, deadlines and click failures. All 612 Node tests pass.

File: scripts/lib/playwright-ui.js [hunk 4]
```diff
@@ -201,7 +205,10 @@ async function clickDownloadsOrOpens(page, locator, options = {}) {
   const downloaded = page.waitForEvent('download', { timeout })
     .then((download) => ({ kind: 'download', url: download.url(), download, page: null }), never);
   const popped = context.waitForEvent('page', { timeout })
-    .then((popup) => ({ kind: 'popup', url: '', download: null, page: popup }), never);
+    .then((popup) => {
+      if (options.recorder) wireStrictPage(popup, label, options.recorder, options);
+      return { kind: 'popup', url: '', download: null, page: popup };
+    }, never);
```
[the remark this answers is not in the document]
Comment (yingbull) [reply scripts/lib/playwright-ui.js]: Fixed in 7a112adfd0 with the same disposable outcome watcher. A completed download removes the popup listener immediately; a later popup cannot inherit its recorder. Click-error and deadline paths also remove both listeners. Full Node suite: 612 passed.

File: src/test/java/io/github/carlos_emr/carlos/commn/model/GroupMembersUnitTest.java [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+import java.lang.reflect.Field;
+import org.junit.jupiter.api.Tag;
+import org.junit.jupiter.api.Test;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+@Tag("unit")
+class GroupMembersUnitTest {
+    @Test
+    void shouldHydrateLegacyNullLocationAsLocal() throws Exception {
```
[the remark this answers is not in the document]
Comment (yingbull) [reply src/test/java/io/github/carlos_emr/carlos/commn/model/GroupMembersUnitTest.java]: Renamed both methods to the repository BDD action_whenContext convention in 7a112adfd0.

File: docs/ui-tests/README.md [hunk 6]
```diff
@@ -6,6 +6,7 @@ Comprehensive UI testing for CARLOS EMR using Playwright MCP (Model Context Prot
 
 - **New to UI testing?** Start with [UI-TEST-PROCESS.md](UI-TEST-PROCESS.md) - Complete testing procedures
 - **Validating the .deb packages?** Use [deb-install-validation.md](deb-install-validation.md) - Build, install into a VM, and run the whole Playwright suite through the nginx + ModSecurity front door
+- **Release workflow results and reproduction?** Read [release-2026.08-workflow-validation.md](release-2026.08-workflow-validation.md) — packaged VM results, new lifecycle checks, and the aggregate findings ticket
```
[the remark this answers is not in the document]
Comment (yingbull) [reply docs/ui-tests/README.md]: Updated the README link label to validation record and reproduction in c5a0302627. The record now distinguishes passed retests from final-package checks still pending.

File: src/main/java/io/github/carlos_emr/carlos/commn/model/GroupMembers.java [hunk 7]
```diff
@@ -90,7 +93,7 @@ public void setFacilityId(int facilityId) {
     }
 
     public int getClinicLocationNo() {
-        return clinicLocationNo;
+        return clinicLocationNo == null ? 0 : clinicLocationNo;
```
[the remark this answers is not in the document]
Comment (yingbull) [reply src/main/java/io/github/carlos_emr/carlos/commn/model/GroupMembers.java:102]: Added JavaDoc documenting persisted NULL → zero and the local-location meaning in c5a0302627. Behavior is unchanged.

File: src/test/java/io/github/carlos_emr/carlos/eform/actions/DisplayImage2ActionUnitTest.java [hunk 8]
```diff
@@ -377,6 +377,50 @@ void shouldThrowSecurityException_whenRequestedImagePathTraversesOutsideAllowedD
...
+            mockRequest = new MockHttpServletRequest(new MockServletContext("file:" + warRoot));
+            mockRequest.setParameter("imagefile", "vaccine-brands.json");
+            servletActionContextMock.when(ServletActionContext::getRequest).thenReturn(mockRequest);
+            action = new DisplayImage2Action();
+            when(mockSecurityInfoManager.hasPrivilege(eq(mockLoggedInInfo), eq("_prevention"), eq("r"), isNull()))
+                    .thenReturn(true);
+        }
+
+        @Test
+        void shouldServeBundledJsonWhenClinicOverrideIsAbsent() throws Exception {
```
[the remark this answers is not in the document]
Comment (yingbull) [reply src/test/java/io/github/carlos_emr/carlos/eform/actions/DisplayImage2ActionUnitTest.java]: Already addressed in 7a112adfd0: all three added methods now use exactly one underscore separating action and condition; the 404 case is shouldReturnNotFound_whenNeitherCatalogueExists.

**Comment by yingbull** [comment]
@coderabbitai review

Addressed the three inline findings and the two outside-diff documentation findings: the validation-record link, nullable getter JavaDoc, BDD test names, and purpose/features/parameters/@since comments in all three touched contact JSPs.

Also fixed Copilot's valid event-race finding across all three popup helpers. Temporary listeners are disposed on success, deadline, and click failure; first-document recording remains synchronous. All 614 Node tests pass. Five original-versus-fixed probes reproduce stale listeners in the prior code and zero in the corrected code. Incorrect dialog-fallback and handler-encoding comments have evidence-based replies.

Final VM validation remains explicitly pending: LXD stalled stopping the isolated validation VM, including force/cancellation, and host administrator intervention is needed. No compilation is running alongside that VM. The PR stays draft until the final package/contact/routing and cleanup retests can complete.

File: src/main/webapp/WEB-INF/jsp/demographic/contact.jsp [hunk 9]
```diff
@@ -46,7 +46,7 @@
 %>
 
 <div id="contact_<carlos:encode value='<%= id %>' context="htmlAttribute"/>">
-    <input type="hidden" name="contact_<carlos:encode value='<%= id %>' context="htmlAttribute"/>.id" id="contact_<carlos:encode value='<%= id %>' context="htmlAttribute"/>.id" value=""/>
+    <input type="hidden" name="contact_<carlos:encode value='<%= id %>' context="htmlAttribute"/>.id" id="contact_<carlos:encode value='<%= id %>' context="htmlAttribute"/>.id" value="0"/>
```
[the remark this answers is not in the document]
Comment (yingbull) [reply src/main/webapp/WEB-INF/jsp/demographic/contact.jsp]: Fixed in bbeb89cbc7: removeContact skips the zero sentinel before DAO lookup for both categories (and the single-ID AJAX path). Review also exposed an existing ClassCastException from casting Object[] to String[]; changed to typed toArray(new String[0]). Three Java regressions cover unsaved IDs, both persisted categories and AJAX response preservation. The UI workflow now discards both kinds of unsaved row before its full lifecycle. Full Node suite remains green; Java CI and final installed-package validation are pending because the host VM shutdown is blocked.

File: src/main/webapp/WEB-INF/jsp/demographic/procontact.jsp [hunk 10]
```diff
@@ -46,7 +46,7 @@
 %>
 
 <div id="procontact_<carlos:encode value='<%= id %>' context="htmlAttribute"/>">
-    <input type="hidden" name="procontact_<carlos:encode value='<%= id %>' context="htmlAttribute"/>.id" id="procontact_<carlos:encode value='<%= id %>' context="htmlAttribute"/>.id" value=""/>
+    <input type="hidden" name="procontact_<carlos:encode value='<%= id %>' context="htmlAttribute"/>.id" id="procontact_<carlos:encode value='<%= id %>' context="htmlAttribute"/>.id" value="0"/>
```
[the remark this answers is not in the document]
Comment (yingbull) [reply src/main/webapp/WEB-INF/jsp/demographic/procontact.jsp]: Fixed in bbeb89cbc7: removeContact skips the zero sentinel before DAO lookup for both categories (and the single-ID AJAX path). Review also exposed an existing ClassCastException from casting Object[] to String[]; changed to typed toArray(new String[0]). Three Java regressions cover unsaved IDs, both persisted categories and AJAX response preservation. The UI workflow now discards both kinds of unsaved row before its full lifecycle. Full Node suite remains green; Java CI and final installed-package validation are pending because the host VM shutdown is blocked.

File: src/main/java/io/github/carlos_emr/carlos/commn/web/Contact2Action.java [hunk 11]
```diff
@@ -421,14 +421,19 @@ public String removeContact() {
...
-                ids = (String[]) arrayListIds.toArray();
+                ids = arrayListIds.toArray(new String[0]);
             }
         }
 
         if (ids != null && ids.length > 0) {
             int contactId;
             for (String id : ids) {
                 contactId = Integer.parseInt(id);
+                // New association rows use zero until first save. Removing one
```
[the remark this answers is not in the document]
Comment (yingbull) [reply src/main/java/io/github/carlos_emr/carlos/commn/web/Contact2Action.java]: Fixed in 766030f016. Contact deletion now requires patient write privilege and matching association ownership, rejects missing/foreign associations, and validates the complete selection before any soft-delete merge. Save and removal both require POST before side effects. Added regressions for write denial, missing context/association, cross-patient IDs, mixed-owner batches and GET/HEAD/PUT/DELETE. Latest Java CI is running; final VM validation remains blocked by the stuck host QEMU process.

File: src/main/java/io/github/carlos_emr/carlos/commn/web/Contact2Action.java [hunk 12]
```diff
@@ -421,14 +421,19 @@ public String removeContact() {
...
+                // in the editor must not look up a nonexistent persisted row.
+                if (contactId == 0) {
+                    continue;
+                }
```
[the remark this answers is not in the document]
Comment (yingbull) [reply src/main/java/io/github/carlos_emr/carlos/commn/web/Contact2Action.java]: Fixed in 766030f016. Contact deletion now requires patient write privilege and matching association ownership, rejects missing/foreign associations, and validates the complete selection before any soft-delete merge. Save and removal both require POST before side effects. Added regressions for write denial, missing context/association, cross-patient IDs, mixed-owner batches and GET/HEAD/PUT/DELETE. Latest Java CI is running; final VM validation remains blocked by the stuck host QEMU process.

**Comment by yingbull** [comment]
@coderabbitai review

Host recovery cleared the stuck VM process. The VM is now stopped and the final full Java/package build is running under a 5 GiB cap, followed by installed-package live validation.

Follow-up fixes include POST-only contact save/removal, write privilege and patient ownership checks, prevalidation of a mixed-owner deletion selection, and negative Java regressions. Personal/professional control IDs are now distinct, controls have labels, and the live workflow asserts ID uniqueness. Sonar's nested-tag parsing issue was removed using the existing safe EL encoding function. All 614 Node regressions and eight real-Chromium helper probes passed before the final package cycle. Please review the follow-up diff.

**Comment by yingbull** [comment]
@coderabbitai review

Addressed the professional-field persistence finding in 2caa03b91f. Also addressed Copilot's save-ownership and Plot-encoding findings. Prevalidation now covers both contact categories and removals before writes; reciprocal writes require permission on both patients and create a separate reverse association instead of reusing/moving the original row.

A discovery audit caught `Contact2ActionTest` excluded by the normal Surefire patterns. Renamed it `Contact2ActionUnitTest`; all 19 focused regressions now actually execute and pass. Full default-suite verification and packaged live retests are underway with the VM stopped during compilation. Node: 614 passed. Please review this follow-up against the PR's actual base `release/2026.08` (not develop).

File: src/main/java/io/github/carlos_emr/carlos/commn/web/Contact2Action.java [hunk 13]
```diff
@@ -197,6 +197,9 @@ public String manage() {
     // FindSecBugs IMPROPER_UNICODE: case-insensitive comparison of an internal/domain value (status/flag/enum/MIME/code); not a security or authorization decision. See docs/static-analysis-workflows.md
     @SuppressFBWarnings(value = "IMPROPER_UNICODE", justification = "case-insensitive comparison of an internal/domain value (status/flag/enum/MIME/code); not a security or authorization decision")
     public String saveManage() {
+        if (!requireContactPost()) {
+            return NONE;
+        }
```
[the remark this answers is not in the document]
Comment (yingbull) [reply src/main/java/io/github/carlos_emr/carlos/commn/web/Contact2Action.java:211]: Fixed in 2caa03b91f. Both contact categories and removals prevalidate every association owner before writes. Reciprocal writes require permission on both patients and create a separate association instead of reusing the original ID. The focused 19-case suite and normal full suite (12,128 tests, zero failures/errors, 51 skips) pass; Surefire XML confirms these contact cases execute after correcting the legacy test filename.

File: src/main/webapp/WEB-INF/jsp/encounter/oscarMeasurements/DisplayHistory.jsp [hunk 14]
```diff
@@ -273,8 +273,9 @@
...
                             </security:oscarSec>
 
-                            <!-- Plot button (same as before) -->
-                            <c:if test="${not empty data.canPlot}">
+                            <%-- The loop-scoped 'data' variable no longer exists here. All rows
+                                 describe this measurement type; use its first row's plot capability. --%>
+                            <c:if test="${not empty measurementsData.measurementsDataVector and not empty measurementsData.measurementsDataVector[0].canPlot}">
                                 <td>
                                     <input type="button" name="Button" class="btn DoNotPrint"
                                         value="<fmt:message key="encounter.oscarMeasurements.displayHistory.plot"/>"
```
[the remark this answers is not in the document]
Comment (yingbull) [reply src/main/webapp/WEB-INF/jsp/encounter/oscarMeasurements/DisplayHistory.jsp:281]: Fixed in 2caa03b91f using forUriComponent(type) followed by forJavaScriptAttribute, matching the existing Measurements.jsp pattern. Final packaged plot retest is pending.

File: scripts/lib/playwright-ui.js [hunk 15]
```diff
@@ -45,6 +45,34 @@ const {
...
+    settled = true;
+    clearTimeout(expire);
+    for (const [emitter, event, listener] of listeners) emitter.off(event, listener);
+  };
+  const promise = new Promise((resolve, reject) => {
+    for (const { emitter, event, accepts = () => true, result } of sources) {
+      const listener = value => {
+        if (settled || !accepts(value)) return;
+        cancel();
+        try { resolve(result(value)); } catch (error) { reject(error); }
```
[the remark this answers is not in the document]
Comment (yingbull) [reply scripts/lib/playwright-ui.js:66]: Fixed in 7a31418299. watchOutcomes retains abandonment cleanup for the observed popup; all three helpers invoke it before propagating a click/event failure, even if the popup won first. Six executable cases fail before and pass after; they also preserve the original error when close fails. A same-tab case preserves its host. All 621 Node regressions pass; three actual Chromium probes independently confirm closure and error/host preservation.

**Comment by yingbull** [comment]
@coderabbitai review

Follow-up in 3d59c528e6: live cleanup audit caught one archived measurement and one chart autosave left behind. Cleanup now deletes and verifies owned support rows before deleting the patient; silent child-delete failure retains the parent and fails. Directory cleanup also verifies zero rows. All seven workflows pass again on the final installed package; a schema-wide audit of all 12 owned patients finds only 60 intentionally retained audit-log rows. Node: 624 passed.

The Master Record's patient-number heading was confirmed to be a plain link to the current document. Such links now count as skipped, never opened coverage; handlers, popups and differing destinations remain audited. Actual Client Lab Label UI response is HTTP 200/PDF MIME/zero bytes, with a caught Jasper template exception; tracked as an open application defect in #3682.

The full normal Java build passed 12,128 tests (zero failures/errors, 51 skips), including all 19 contact tests. Final package health passes; reinstall applies zero migrations. Administration and final login/restoration checks are finishing. Please review the final test/cleanup changes against release/2026.08.

File: src/main/java/io/github/carlos_emr/carlos/commn/web/Contact2Action.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (yingbull) [reply src/main/java/io/github/carlos_emr/carlos/commn/web/Contact2Action.java:211]: Final packaged contact workflow passes all five steps, including personal edits/removal and professional false-to-true consent/status round-trips. All 19 Java action regressions are confirmed in the passing default full suite. The final cleanup audit leaves only intentional audit logs.

File: src/main/webapp/WEB-INF/jsp/encounter/oscarMeasurements/DisplayHistory.jsp (hunk 14 above)
[the remark this answers is not in the document]
Comment (yingbull) [reply src/main/webapp/WEB-INF/jsp/encounter/oscarMeasurements/DisplayHistory.jsp:281]: Final installed-package measurement workflow passes: correct dates/values, selective deletion with archived value retained, and Plot returning HTTP 200 with real PNG bytes. Encoding matches the established URI-component plus JavaScript-attribute pattern.

**Comment by yingbull** [comment]
@coderabbitai review

Please review 963063f241, especially reciprocal contact authorization with an omitted type, saves that do not write the related chart, and reverse-row type/SDM/EC persistence. Five new regression failures are reproduced against the previous implementation; corrected full Java and packaged VM validation are running.

**Comment by yingbull** [comment]
@coderabbitai review

Follow-up 73ce9d8d95 addresses the static-analysis findings: shared parameter constants, unshadowed association variables, and a smaller reciprocal-planning helper with one loop continuation. The full normal Java suite for the functional fix passes 12,134 tests with zero failures/errors and 51 skips, including all 25 contact cases. Final refactor/package and VM before/after checks are next.

**Comment by yingbull** [comment]
@codex review

Please recheck the reciprocal-write fix and follow-up refactor in 963063f241 and 73ce9d8d95, including omitted/padded types, no-write saves with restricted related-patient access, and reciprocal SDM/EC flags. The expanded regressions reproduce the old failures and all 25 pass in the default full Java suite.

**Comment by yingbull** [comment]
@coderabbitai review

Final follow-up: 555cb944ef keeps the related-patient fixture within the actual VARCHAR(30) surname limit; its old-package negative test still fails on the intended type/SDM/EC assertion and now cleans up successfully. The corrected DEB passes all seven workflows, including the new reciprocal-contact step, plus all 14 login checks with credential restoration verified. Final cleanup audit: 12 owned patients, only 49 retained audit-log rows.

362a662978 finishes the three field-suffix constants and corrects the helper ID-contract documentation. All 25 action tests pass again; normalized javap instructions and invokedynamic recipes are identical to the VM-tested 73ce9d8d95 action. No Flyway/schema changes; origin release/2026.08 remains fcb14db3. VM stopped, memory guards recorded no swap or interventions. Please review the final small follow-ups.

File: src/main/java/io/github/carlos_emr/carlos/commn/web/Contact2Action.java [hunk 16]
```diff
@@ -379,64 +379,76 @@ private String getReverseRole(String roleName, int targetDemographicNo) {
...
+            String field = prefix + row;
+            String id = request.getParameter(field + ".id");
+            if (id == null) continue;
+            int associationId = Integer.parseInt(id);
+            if (associationId != 0) requireOwnedContact(associationId, demographicNo);
+            String contactId = request.getParameter(field + ".contactId");
+            if ("contact_".equals(prefix) && "1".equals(request.getParameter(field + ".type"))
+                    && StringUtils.isNotBlank(contactId) && !"0".equals(contactId)
+                    && !securityInfoManager.hasPrivilege(loggedInInfo, "_demographic", "w", contactId)) {
+                throw new SecurityException("missing write permission for reciprocal contact patient");
```
[the remark this answers is not in the document]
Comment (yingbull) [reply src/main/java/io/github/carlos_emr/carlos/commn/web/Contact2Action.java]: Fixed in 963063f241 and simplified in 73ce9d8d95. Validation derives an omitted type from the owned persisted association, plans only actual reverse-row creations, and requires related-patient write access only for that plan before any mutation. Existing reverse links and unmapped roles no longer block unrelated/current-patient saves. New tests cover both omitted and numeric 01 types, restricted related-patient access, no partial writes and successful reciprocal persistence. All 25 cases pass in the normal 12,134-test suite and the final focused run. The old installed package fails the new reciprocal UI assertion; the corrected DEB passes all seven workflows, including correct reverse type/flags and duplicate prevention. Both automated follow-up reviewers found no new major issues in 73ce9d8d95.

**Comment by yingbull** [comment]
Reviewed the suppressed findings as well as all inline threads:

- Fixed browser-teardown cleanup in `758801a532`. Negative test: 7/8 pass before the fix, with the owned-parent-deletion assertion failing. Afterward all 626 Node tests pass. Browser-close failure remains visible, successful database cleanup removes the parent, and child failures still retain it. This changes the harness only; VM-tested application runtime is unchanged.
- The SOAP fixture is intentionally Object-valued. The recorded negative run against the original XML fails in Spring `resolveMultipleBeanMap` → `autowireByType`, trying to instantiate `requestDependentAction` for the `properties` map. Thus it already catches restoration of `autowire="byType"`; the fixed wiring passes. Added an explanatory test comment.
- Popup save has no asynchronous yield between `isClosed()` and subscribing to `close`. A close processed before the check is detected; a later event cannot interleave those synchronous calls. No demonstrated missed-event race here.
- Numeric parsing before authorization is inherited from the release base, and no writes occur before authorization. Cataloged its malformed-input error handling as source-confirmed finding 36 in #3682; not claiming a live response reproduction or permission bypass.

@coderabbitai Please review the final cleanup follow-up `758801a532` and its two regression cases.

File: src/test/java/io/github/carlos_emr/carlos/webserv/AuthenticationInterceptorWiringUnitTest.java [hunk 17]
```diff
@@ -0,0 +1,83 @@
...
+        }
+
+        AtomicInteger requestActionCreations = new AtomicInteger();
+        OscarUsernameTokenValidator validator = mock(OscarUsernameTokenValidator.class);
+        try (GenericApplicationContext context = new GenericApplicationContext()) {
+            AnnotationConfigUtils.registerAnnotationConfigProcessors(context);
+            context.getBeanFactory().registerSingleton("oscarUsernameTokenValidator", validator);
+            RootBeanDefinition action = new RootBeanDefinition(Object.class, () -> {
+                requestActionCreations.incrementAndGet();
+                throw new IllegalStateException("Request action cannot be created outside an HTTP request");
```
[the remark this answers is not in the document]
Comment (yingbull) [reply src/test/java/io/github/carlos_emr/carlos/webserv/AuthenticationInterceptorWiringUnitTest.java:67]: This is contradicted by the executable negative run against the original XML. The test fails with UnsatisfiedDependencyException on bean property `properties`, caused by `requestDependentAction` creation. Its stack includes `DefaultListableBeanFactory.resolveMultipleBeanMap` → `autowireByType`, and the prototype supplier throws "Request action cannot be created outside an HTTP request". Spring populates the Map with Object-valued candidates; those candidates need not themselves implement Map. The fixed XML passes. The fixture is therefore intentional and already guards this regression; 758801a532 adds an explanatory comment. CodeRabbit independently reviewed that reasoning. Negative evidence is summarized in docs/ui-tests/release-2026.08-workflow-validation.md; preserving the meaningful scenario rather than changing its candidate type.

**Comment by yingbull** [comment]
@coderabbitai Please review `f94f0d158e`: two additional suppressed findings were valid and are now fixed. The required administration panel can no longer fall back to shell text; real full-document navigation remains supported and validated, while hash changes do not bypass the panel requirement. Prevention creation/replacement reads all active rows and requires exactly one instead of taking the scalar SQL helper's first row. Missing-panel negative test reproduces the old false pass; all 635 Node tests pass, including missing/duplicate/invalid prevention rows and navigation/error-page cases. Targeted VM/Chromium revalidation is running with memory guards and no host compilation. The SOAP inline suggestion is addressed separately with its recorded old-XML failure stack.

**Comment by yingbull** [comment]
@coderabbitai Please assess the remaining type-immutability suggestion (inline 4029452219) against the current authorization and persistence contract. Existing-row `.type` is currently an explicit supported input to `linkContactToDemographic`; omitted type preserves persisted type. Both reciprocal planning and persistence use the same effective submitted/stored type. Source-patient write permission and association ownership are checked before any writes; target write permission is checked only when a reverse patient row is planned. Is there a reproducible unauthorized write or inconsistent/silent result here, or is making the type immutable an additional API/UI contract restriction? Please trace both 1→2 and 2→1 and consider whether delete/recreate already permits the same authorized source change. The editor disables the selector for existing rows, so stricter server enforcement could be a hardening follow-up; please assess independently.

The latest suppressed claims also need the actual call site checked: `findNewReciprocalRole(..., demographicNo)` passes the already-authorized SOURCE patient's ID to `getReverseRole`, not `contactId`. The source patient's sex determines their reverse role, so the comment asserting an unauthorized target-demographic read appears to confuse the helper parameter name with its argument. The prevention cardinality issue is fixed in f94f0d158e by querying all active rows and requiring exactly one; its new negative tests and live VM check pass.

File: src/main/java/io/github/carlos_emr/carlos/commn/web/Contact2Action.java [hunk 18]
```diff
@@ -379,64 +377,96 @@ private String getReverseRole(String roleName, int targetDemographicNo) {
...
+    private String findNewReciprocalRole(String field, DemographicContact existing, int demographicNo) {
+        String contactId = request.getParameter(field + CONTACT_ID_SUFFIX);
+        if (StringUtils.isBlank(contactId) || "0".equals(contactId)) return null;
+
+        // Existing rows disable their type selector, so browsers omit it.
+        // Match the persisted-type fallback used by linkContactToDemographic.
+        String submittedType = request.getParameter(field + CONTACT_TYPE_SUFFIX);
+        int effectiveType = existing != null ? existing.getType() : DemographicContact.TYPE_PROVIDER;
+        if (submittedType != null) effectiveType = Integer.parseInt(submittedType);
+        if (effectiveType != DemographicContact.TYPE_DEMOGRAPHIC
```
[the remark this answers is not in the document]
Comment (yingbull) [reply src/main/java/io/github/carlos_emr/carlos/commn/web/Contact2Action.java:452]: Fixed in 5542019d94. Existing association types are now preserved in both reciprocal planning and persistence; only new rows parse a submitted type. Added five regressions covering both conversion directions, denial before any mutation, correct reciprocal creation, and professional types including malformed submissions. Before fix: 30 cases, 3 failures + 2 errors; after fix: all 30 pass. Rebuilt WAR/DEB and verified installed hashes. Old installed package fails an owned-fixture request-tampering probe; rebuilt package passes all six normal contact steps and the probe with both saves tampered, preserving type/reverse association and avoiding duplicates. Cleanup: six owned patients, only 26 retained audit rows; zero migrations; health passes, VM stopped and original mounts restored. Validation record updated. The getReverseRole parameter is now sourceDemographicNo; no unauthorized target-demographic read occurs.

**Comment by yingbull** [comment]
@coderabbitai review

Please verify final consistency fix `5542019d94`: existing association types remain persisted in both reciprocal planning and save; new rows still use the selected type. Five negative cases fail the preceding code, all 30 pass the corrected build. The old installed DEB fails a controlled owned-contact `.type=2` request; the rebuilt DEB passes the normal six-step contact workflow and two tampered saves, with installed hashes/cleanup/zero migrations verified. VM stopped, mounts restored. This explicitly addresses inline 4029452219 and your consistency assessment. The earlier category-conversion suggestion is retained as source-only follow-up finding 38 in #3682; it does not block the requested UI tests and was not reproduced as an authorization bypass. Final full CI running on this head.
